app/src/main/java/com/ikarabulut/GameSettings.java

```diff
@@ -32,35 +34,93 @@ public void getAllSettings() {
 
     public void setEvolutions() {
         display.evolutionsPrompt();
-        evolutions = Integer.parseInt(input.nextLine());
+        try {
+            evolutions = Integer.parseInt(input.nextLine());
+        } catch (Exception e) {
+            System.err.print("An error occurred, please enter an integer. Try again: ");
+            setEvolutions();
+        }
     }
 
+
     public void setRows() {
         display.rowsPrompt();
-        rows = Integer.parseInt(input.nextLine());
+        try {
+            intSelection = Integer.parseInt(input.nextLine());
+        } catch (Exception ex) {
+            System.err.print("Your input is invalid. Please try again: ");
+            setRows();
+        } finally {
```

Comment:
I see how you're doing retries now.  A better technique for doing this is to keep a retry counter.  Here is an example:

```
    public void setRows() {
        display.rowsPrompt();
        retries = 3;

        while (retries > 0) {
            try {
                selection = Integer.parseInt(input.nextLine());
            } catch (NumberFormatException ex) {
                System.err.print("Your input is invalid please try again: ");
                retries = retries - 1;
            }
        }

        if (selection == null) {
            System.err.print("You have failed to choose a valid number of rows.");
            // What happens here? Raise an application-specific exception like "InvalidInputException" that the caller should handle?
        }
    }
```

Reply:
Thanks for this technique to allow the user to retry. I missed the `if` conditional at the bottom which is within the `while` loop. Also I found out I need to add a `return` after setting the selection to the user input, or else the program will keep taking the input since the counter is above 0